public/app/core/logs_model.ts

```diff
@@ -326,3 +325,141 @@ export function makeSeriesForLogs(rows: LogRowModel[], intervalMs: number): Time
     };
   });
 }
+
+export function seriesDataToLogsModel(seriesData: SeriesData[], intervalMs: number): LogsModel {
+  const metricSeries: SeriesData[] = [];
+  const logSeries: SeriesData[] = [];
+
+  for (const series of seriesData) {
+    if (series.name !== 'logs') {
+      metricSeries.push(series);
+      continue;
+    }
+
+    if (getFirstTimeField(series) === -1) {
+      continue;
+    }
+
+    if (!hasFieldNamed(series, 'message')) {
+      continue;
+    }
+
+    logSeries.push(series);
+  }
+
+  const logsModel = logSeriesToLogsModel(logSeries);
+  if (logsModel) {
+    if (metricSeries.length === 0) {
+      logsModel.series = makeSeriesForLogs(logsModel.rows, intervalMs);
+    } else {
+      logsModel.series = [];
+      for (const series of metricSeries) {
+        logsModel.series.push(toLegacyResponseData(series) as TimeSeries);
+      }
+    }
+
+    return logsModel;
+  }
+
+  return undefined;
+}
+
+export function logSeriesToLogsModel(logSeries: SeriesData[]): LogsModel {
+  if (logSeries.length === 0) {
+    return undefined;
+  }
+
+  // Unique model identifier
+  const id = logSeries
+    .map(series => {
+      if (!series.labels) {
+        return '';
+      }
+
+      return Object.keys(series.labels)
+        .map(key => `key=${series.labels[key]}`)
+        .join();
+    })
+    .join();
+
+  // Find unique labels for each series
+  const commonLabels = findCommonLabels(logSeries.map(series => series.labels));
+  logSeries = logSeries.map(series => ({
+    ...series,
+    uniqueLabels: findUniqueLabels(series.labels, commonLabels),
+  }));
+
+  const uniqueLabelsPerSeries: Labels[] = [];
+  for (let n = 0; n < logSeries.length; n++) {
+    const series = logSeries[n];
+    uniqueLabelsPerSeries.push(findUniqueLabels(series.labels, commonLabels));
+  }
+
+  // Merge series entries into single list of log rows
+  const sortedRows: LogRowModel[] = _.chain(logSeries)
+    .reduce(
+      (acc: LogRowModel[], series: SeriesData, index: number) => [
+        ...acc,
+        ...series.rows.map(row => processLogSeriesRow(series, row, uniqueLabelsPerSeries[index])),
+      ],
+      []
+    )
+    .sortBy('timestamp')
+    .reverse()
+    .value();
+
+  const hasUniqueLabels = sortedRows && sortedRows.some(row => Object.keys(row.uniqueLabels).length > 0);
```

Comment:
doing this in the main row loop would be preferable instead of looping in worst case a lot rows again (via callback fn). also why the sortedRows null check? how can it be null?

Reply:
Have tried to fix this. Please review